File: services/simulator/src/money.ts
// Money is a BIGINT count of minor units plus a currency code. There is no
// float anywhere in this codebase, and no bare number that means "dollars".

export type Currency = "USD" | "EUR" | "GBP" | "JPY";

// Not every currency has cents. JPY has no minor unit at all, which is why
// this is a table and not the number two: ¥5,000 stored as 5000 minor units
// renders as ¥5,000, and a formatter that assumes two decimals renders ¥50.
const MINOR_UNITS: Record<Currency, number> = { USD: 2, EUR: 2, GBP: 2, JPY: 0 };

export function toMinor(major: number, currency: Currency): number {
  const scale = 10 ** MINOR_UNITS[currency];
  // Round once, at the boundary, and never again.
  return Math.round(major * scale);
}

export function formatMinor(minor: number | bigint, currency: Currency): string {
  const value = typeof minor === "bigint" ? minor : BigInt(Math.trunc(minor));
  const scale = BigInt(10 ** MINOR_UNITS[currency]);
  const sign = value < 0n ? "-" : "";
  const abs = value < 0n ? -value : value;
  const whole = abs / scale;
  const digits = MINOR_UNITS[currency];
  if (digits === 0) {
    // No fractional part, and therefore no separator. Without this the yen
    // amount comes out as "5,000. JPY".
    return `${sign}${whole.toLocaleString("en-US")} ${currency}`;
  }
  const frac = (abs % scale).toString().padStart(digits, "0");
  return `${sign}${whole.toLocaleString("en-US")}.${frac} ${currency}`;
}

File: internal/agent/money.go
package agent

import (
	"fmt"
	"strings"
)

// Money crosses into a prompt formatted, never as raw minor units.
//
// This file exists because of a letter that said $579.99 about a $57.99
// dispute. The record handed the model amount_minor: 5799 and the prompt told
// it to copy amounts from the record digit for digit - which is exactly what it
// did. Another draft copied 2790 straight through for a charge of $27.90.
//
// The rule everywhere else in this system is that money is divided exactly
// once, at the formatter: the ledger stores minor units, the dashboard divides
// at the edge, and the API never sends a float. The prompt was the one boundary
// that got an undivided integer and a human-language instruction to transcribe
// it. A model doing arithmetic on the way to a card network is a bug in the
// prompt, not in the model.
//
// Not every currency has two decimal places, which is why this is a table and
// not the number 100.
var minorDigits = map[string]int{"JPY": 0, "KRW": 0, "ISK": 0}

func digitsFor(currency string) int {
	if d, ok := minorDigits[strings.ToUpper(currency)]; ok {
		return d
	}
	return 2
}

// FormatMinor renders an amount the way a letter should state it.
func FormatMinor(amountMinor int64, currency string) string {
	digits := digitsFor(currency)
	if digits == 0 {
		return fmt.Sprintf("%s %s", group(amountMinor), strings.ToUpper(currency))
	}

	scale := int64(1)
	for range digits {
		scale *= 10
	}
	whole, frac := amountMinor/scale, amountMinor%scale
	if frac < 0 {
		frac = -frac
	}
	return fmt.Sprintf("%s.%0*d %s", group(whole), digits, frac, strings.ToUpper(currency))
}

// group inserts thousands separators. A four-figure amount written without one
// is the shape a misplaced decimal takes, and the letters this produces are
// read by somebody deciding whether to reverse a charge.
func group(n int64) string {
	sign := ""
	if n < 0 {
		sign, n = "-", -n
	}
	digits := fmt.Sprintf("%d", n)

	var b strings.Builder
	for i, r := range digits {
		if i > 0 && (len(digits)-i)%3 == 0 {
			b.WriteByte(',')
		}
		b.WriteRune(r)
	}
	return sign + b.String()
}
